eth/downloader/bor_fetchers_concurrent.go

```diff
@@ -221,7 +230,20 @@ func (d *Downloader) concurrentFetch(queue typedQueue, beaconMode bool) error {
 			}
 			// Make sure that we have peers available for fetching. If all peers have been tried
 			// and all failed throw an error
-			if !progressed && !throttled && len(pending) == 0 && len(idles) == d.peers.Len() && queued > 0 && !beaconMode {
+			var capablePeers int
+			if isWitnessQueue {
+				// For witness fetching, only count peers that support the witness protocol
+				for _, peer := range d.peers.AllPeers() {
+					if peer.peer.SupportsWitness() {
+						capablePeers++
+					}
+				}
+			} else {
+				// For other queues, all peers are capable
+				capablePeers = d.peers.Len()
+			}
+
+			if !progressed && !throttled && len(pending) == 0 && len(idles) == capablePeers && queued > 0 && !beaconMode {
```

Comment:
This check does `len(idles) == capablePeers`. Even if there's a non-overlapping set of idle peers and peers being able to support witness protocol, won't this check return an error in those cases? 

E.g. if total peers are 20, 10 are idle and 10 support witness. But, we don't know if those 10 peers are same ones so it would be unfair to return an error in such case.

Reply:
Ah okay didn't see that as it wasn't in the diff. Thanks!